File: airbyte-server/src/main/java/io/airbyte/server/apis/WebBackendApiController.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.server.apis;

import static io.airbyte.commons.auth.AuthRoleConstants.AUTHENTICATED_USER;
import static io.airbyte.commons.auth.AuthRoleConstants.ORGANIZATION_EDITOR;
import static io.airbyte.commons.auth.AuthRoleConstants.ORGANIZATION_READER;
import static io.airbyte.commons.auth.AuthRoleConstants.WORKSPACE_EDITOR;
import static io.airbyte.commons.auth.AuthRoleConstants.WORKSPACE_READER;

import io.airbyte.api.generated.WebBackendApi;
import io.airbyte.api.model.generated.ConnectionIdRequestBody;
import io.airbyte.api.model.generated.ConnectionStateType;
import io.airbyte.api.model.generated.PermissionType;
import io.airbyte.api.model.generated.WebBackendCheckUpdatesRead;
import io.airbyte.api.model.generated.WebBackendConnectionCreate;
import io.airbyte.api.model.generated.WebBackendConnectionListRequestBody;
import io.airbyte.api.model.generated.WebBackendConnectionRead;
import io.airbyte.api.model.generated.WebBackendConnectionReadList;
import io.airbyte.api.model.generated.WebBackendConnectionRequestBody;
import io.airbyte.api.model.generated.WebBackendConnectionUpdate;
import io.airbyte.api.model.generated.WebBackendCronExpressionDescription;
import io.airbyte.api.model.generated.WebBackendDescribeCronExpressionRequestBody;
import io.airbyte.api.model.generated.WebBackendGeographiesListResult;
import io.airbyte.api.model.generated.WebBackendValidateMappersRequestBody;
import io.airbyte.api.model.generated.WebBackendValidateMappersResponse;
import io.airbyte.api.model.generated.WebBackendWorkspaceState;
import io.airbyte.api.model.generated.WebBackendWorkspaceStateResult;
import io.airbyte.commons.lang.MoreBooleans;
import io.airbyte.commons.server.authorization.ApiAuthorizationHelper;
import io.airbyte.commons.server.authorization.Scope;
import io.airbyte.commons.server.handlers.WebBackendCheckUpdatesHandler;
import io.airbyte.commons.server.handlers.WebBackendConnectionsHandler;
import io.airbyte.commons.server.handlers.WebBackendGeographiesHandler;
import io.airbyte.commons.server.scheduling.AirbyteTaskExecutors;
import io.airbyte.commons.server.support.CurrentUserService;
import io.airbyte.metrics.lib.TracingHelper;
import io.airbyte.server.handlers.WebBackendCronExpressionHandler;
import io.airbyte.server.handlers.WebBackendMappersHandler;
import io.micronaut.http.annotation.Body;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Post;
import io.micronaut.scheduling.annotation.ExecuteOn;
import io.micronaut.security.annotation.Secured;
import io.micronaut.security.rules.SecurityRule;
import java.util.Set;

@Controller("/api/v1/web_backend")
@Secured(SecurityRule.IS_AUTHENTICATED)
public class WebBackendApiController implements WebBackendApi {

  private final WebBackendConnectionsHandler webBackendConnectionsHandler;
  private final WebBackendGeographiesHandler webBackendGeographiesHandler;
  private final WebBackendCheckUpdatesHandler webBackendCheckUpdatesHandler;
  private final WebBackendCronExpressionHandler webBackendCronExpressionHandler;
  private final WebBackendMappersHandler webBackendMappersHandler;
  private final ApiAuthorizationHelper apiAuthorizationHelper;
  private final CurrentUserService currentUserService;

  public WebBackendApiController(final WebBackendConnectionsHandler webBackendConnectionsHandler,
                                 final WebBackendGeographiesHandler webBackendGeographiesHandler,
                                 final WebBackendCheckUpdatesHandler webBackendCheckUpdatesHandler,
                                 final WebBackendCronExpressionHandler webBackendCronExpressionHandler,
                                 final WebBackendMappersHandler webBackendMappersHandler,
                                 final ApiAuthorizationHelper apiAuthorizationHelper,
                                 final CurrentUserService currentUserService) {
    this.webBackendConnectionsHandler = webBackendConnectionsHandler;
    this.webBackendGeographiesHandler = webBackendGeographiesHandler;
    this.webBackendCheckUpdatesHandler = webBackendCheckUpdatesHandler;
    this.webBackendCronExpressionHandler = webBackendCronExpressionHandler;
    this.webBackendMappersHandler = webBackendMappersHandler;
    this.apiAuthorizationHelper = apiAuthorizationHelper;
    this.currentUserService = currentUserService;
  }

  @Post("/state/get_type")
  @Secured({WORKSPACE_READER, ORGANIZATION_READER})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public ConnectionStateType getStateType(@Body final ConnectionIdRequestBody connectionIdRequestBody) {
    return ApiHelper.execute(() -> {
      TracingHelper.addConnection(connectionIdRequestBody.getConnectionId());
      return webBackendConnectionsHandler.getStateType(connectionIdRequestBody);
    });
  }

  @Post("/check_updates")
  @Secured({AUTHENTICATED_USER})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public WebBackendCheckUpdatesRead webBackendCheckUpdates() {
    return ApiHelper.execute(webBackendCheckUpdatesHandler::checkUpdates);
  }

  @Post("/connections/create")
  @Secured({WORKSPACE_EDITOR, ORGANIZATION_EDITOR})
  @ExecuteOn(AirbyteTaskExecutors.SCHEDULER)
  @Override
  public WebBackendConnectionRead webBackendCreateConnection(@Body final WebBackendConnectionCreate webBackendConnectionCreate) {
    return ApiHelper.execute(() -> {
      TracingHelper.addSourceDestination(webBackendConnectionCreate.getSourceId(), webBackendConnectionCreate.getDestinationId());
      return webBackendConnectionsHandler.webBackendCreateConnection(webBackendConnectionCreate);
    });
  }

  @Post("/connections/get")
  @Secured({WORKSPACE_READER, ORGANIZATION_READER})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public WebBackendConnectionRead webBackendGetConnection(@Body final WebBackendConnectionRequestBody webBackendConnectionRequestBody) {
    return ApiHelper.execute(() -> {
      TracingHelper.addConnection(webBackendConnectionRequestBody.getConnectionId());
      if (MoreBooleans.isTruthy(webBackendConnectionRequestBody.getWithRefreshedCatalog())) {
        // only allow refresh catalog if the user is at least a workspace editor or
        // organization editor for the connection's workspace
        apiAuthorizationHelper.checkWorkspacesPermissions(
            webBackendConnectionRequestBody.getConnectionId().toString(),
            Scope.CONNECTION,
            currentUserService.getCurrentUser().getUserId(),
            Set.of(PermissionType.WORKSPACE_EDITOR, PermissionType.ORGANIZATION_EDITOR));
      }
      return webBackendConnectionsHandler.webBackendGetConnection(webBackendConnectionRequestBody);
    });
  }

  @Post("/workspace/state")
  @Secured({WORKSPACE_READER, ORGANIZATION_READER})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public WebBackendWorkspaceStateResult webBackendGetWorkspaceState(@Body final WebBackendWorkspaceState webBackendWorkspaceState) {
    return ApiHelper.execute(() -> {
      TracingHelper.addWorkspace(webBackendWorkspaceState.getWorkspaceId());
      return webBackendConnectionsHandler.getWorkspaceState(webBackendWorkspaceState);
    });
  }

  @SuppressWarnings("LineLength")
  @Post("/connections/list")
  @Secured({WORKSPACE_READER, ORGANIZATION_READER})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public WebBackendConnectionReadList webBackendListConnectionsForWorkspace(@Body final WebBackendConnectionListRequestBody webBackendConnectionListRequestBody) {
    return ApiHelper.execute(() -> {
      TracingHelper.addWorkspace(webBackendConnectionListRequestBody.getWorkspaceId());
      return webBackendConnectionsHandler.webBackendListConnectionsForWorkspace(webBackendConnectionListRequestBody);
    });
  }

  @Post("/geographies/list")
  @Secured({AUTHENTICATED_USER})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public WebBackendGeographiesListResult webBackendListGeographies() {
    return ApiHelper.execute(webBackendGeographiesHandler::listGeographiesOSS);
  }

  @Post("/connections/update")
  @Secured({WORKSPACE_EDITOR, ORGANIZATION_EDITOR})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public WebBackendConnectionRead webBackendUpdateConnection(@Body final WebBackendConnectionUpdate webBackendConnectionUpdate) {
    return ApiHelper.execute(() -> {
      TracingHelper.addConnection(webBackendConnectionUpdate.getConnectionId());
      return webBackendConnectionsHandler.webBackendUpdateConnection(webBackendConnectionUpdate);
    });
  }

  @SuppressWarnings("LineLength")
  @Post("/connections/mappers/validate")
  @Secured({WORKSPACE_EDITOR, ORGANIZATION_EDITOR})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public WebBackendValidateMappersResponse webBackendValidateMappers(@Body final WebBackendValidateMappersRequestBody webBackendValidateMappersRequestBody) {
    return ApiHelper.execute(() -> webBackendMappersHandler.validateMappers(webBackendValidateMappersRequestBody));
  }

  @Post("/describe_cron_expression")
  @Secured({AUTHENTICATED_USER})
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @Override
  public WebBackendCronExpressionDescription webBackendDescribeCronExpression(@Body final WebBackendDescribeCronExpressionRequestBody body) {
    return ApiHelper.execute(() -> webBackendCronExpressionHandler.describeCronExpression(body));
  }

}

File: airbyte-server/src/main/kotlin/io/airbyte/server/apis/controllers/WebBackendApiController.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.server.apis.controllers

import io.airbyte.api.generated.WebBackendApi
import io.airbyte.api.model.generated.ConnectionIdRequestBody
import io.airbyte.api.model.generated.ConnectionStateType
import io.airbyte.api.model.generated.WebBackendCheckUpdatesRead
import io.airbyte.api.model.generated.WebBackendConnectionCreate
import io.airbyte.api.model.generated.WebBackendConnectionListRequestBody
import io.airbyte.api.model.generated.WebBackendConnectionRead
import io.airbyte.api.model.generated.WebBackendConnectionReadList
import io.airbyte.api.model.generated.WebBackendConnectionRequestBody
import io.airbyte.api.model.generated.WebBackendConnectionUpdate
import io.airbyte.api.model.generated.WebBackendCronExpressionDescription
import io.airbyte.api.model.generated.WebBackendDescribeCronExpressionRequestBody
import io.airbyte.api.model.generated.WebBackendValidateMappersRequestBody
import io.airbyte.api.model.generated.WebBackendValidateMappersResponse
import io.airbyte.api.model.generated.WebBackendWorkspaceState
import io.airbyte.api.model.generated.WebBackendWorkspaceStateResult
import io.airbyte.api.model.generated.WebappConfigResponse
import io.airbyte.commons.annotation.AuditLogging
import io.airbyte.commons.annotation.AuditLoggingProvider
import io.airbyte.commons.auth.AuthRoleConstants
import io.airbyte.commons.lang.MoreBooleans
import io.airbyte.commons.server.authorization.RoleResolver
import io.airbyte.commons.server.handlers.WebBackendCheckUpdatesHandler
import io.airbyte.commons.server.handlers.WebBackendConnectionsHandler
import io.airbyte.commons.server.scheduling.AirbyteTaskExecutors
import io.airbyte.commons.server.support.AuthenticationId
import io.airbyte.metrics.lib.TracingHelper
import io.airbyte.server.apis.execute
import io.airbyte.server.handlers.WebBackendCronExpressionHandler
import io.airbyte.server.handlers.WebBackendMappersHandler
import io.micronaut.context.annotation.ConfigurationProperties
import io.micronaut.context.annotation.Value
import io.micronaut.http.annotation.Body
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Get
import io.micronaut.http.annotation.Post
import io.micronaut.scheduling.annotation.ExecuteOn
import io.micronaut.security.annotation.Secured
import io.micronaut.security.rules.SecurityRule
import java.util.concurrent.Callable

@Controller("/api/v1/web_backend")
@Secured(SecurityRule.IS_AUTHENTICATED)
open class WebBackendApiController(
  private val webBackendConnectionsHandler: WebBackendConnectionsHandler,
  private val webBackendCheckUpdatesHandler: WebBackendCheckUpdatesHandler,
  private val webBackendCronExpressionHandler: WebBackendCronExpressionHandler,
  private val webBackendMappersHandler: WebBackendMappersHandler,
  private val webappConfig: WebappConfig,
  private val roleResolver: RoleResolver,
) : WebBackendApi {
  @Post("/state/get_type")
  @Secured(AuthRoleConstants.WORKSPACE_READER, AuthRoleConstants.ORGANIZATION_READER)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun getStateType(
    @Body connectionIdRequestBody: ConnectionIdRequestBody,
  ): ConnectionStateType? =
    execute {
      TracingHelper.addConnection(connectionIdRequestBody.connectionId)
      webBackendConnectionsHandler.getStateType(connectionIdRequestBody)
    }

  @Post("/check_updates")
  @Secured(AuthRoleConstants.AUTHENTICATED_USER)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun webBackendCheckUpdates(): WebBackendCheckUpdatesRead? = execute(Callable { webBackendCheckUpdatesHandler.checkUpdates() })

  @Post("/connections/create")
  @Secured(AuthRoleConstants.WORKSPACE_EDITOR, AuthRoleConstants.ORGANIZATION_EDITOR)
  @ExecuteOn(AirbyteTaskExecutors.SCHEDULER)
  @AuditLogging(provider = AuditLoggingProvider.ONLY_ACTOR)
  override fun webBackendCreateConnection(
    @Body webBackendConnectionCreate: WebBackendConnectionCreate,
  ): WebBackendConnectionRead? =
    execute {
      TracingHelper.addSourceDestination(webBackendConnectionCreate.sourceId, webBackendConnectionCreate.destinationId)
      webBackendConnectionsHandler.webBackendCreateConnection(webBackendConnectionCreate)
    }

  @Post("/connections/get")
  @Secured(AuthRoleConstants.WORKSPACE_READER, AuthRoleConstants.ORGANIZATION_READER)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun webBackendGetConnection(
    @Body webBackendConnectionRequestBody: WebBackendConnectionRequestBody,
  ): WebBackendConnectionRead? =
    execute {
      TracingHelper.addConnection(webBackendConnectionRequestBody.connectionId)
      if (MoreBooleans.isTruthy(webBackendConnectionRequestBody.withRefreshedCatalog)) {
        // only allow refresh catalog if the user is at least a workspace editor or
        // organization editor for the connection's workspace
        roleResolver
          .Request()
          .withCurrentUser()
          .withRef(AuthenticationId.CONNECTION_ID, webBackendConnectionRequestBody.connectionId.toString())
          .requireRole(AuthRoleConstants.WORKSPACE_EDITOR)
      }
      webBackendConnectionsHandler.webBackendGetConnection(webBackendConnectionRequestBody)
    }

  @Post("/workspace/state")
  @Secured(AuthRoleConstants.WORKSPACE_READER, AuthRoleConstants.ORGANIZATION_READER)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun webBackendGetWorkspaceState(
    @Body webBackendWorkspaceState: WebBackendWorkspaceState,
  ): WebBackendWorkspaceStateResult? =
    execute {
      TracingHelper.addWorkspace(webBackendWorkspaceState.workspaceId)
      webBackendConnectionsHandler.getWorkspaceState(webBackendWorkspaceState)
    }

  @Post("/connections/list")
  @Secured(AuthRoleConstants.WORKSPACE_READER, AuthRoleConstants.ORGANIZATION_READER)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun webBackendListConnectionsForWorkspace(
    @Body webBackendConnectionListRequestBody: WebBackendConnectionListRequestBody,
  ): WebBackendConnectionReadList? =
    execute {
      TracingHelper.addWorkspace(webBackendConnectionListRequestBody.workspaceId)
      webBackendConnectionsHandler.webBackendListConnectionsForWorkspace(webBackendConnectionListRequestBody)
    }

  @Post("/connections/update")
  @Secured(AuthRoleConstants.WORKSPACE_EDITOR, AuthRoleConstants.ORGANIZATION_EDITOR)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  @AuditLogging(provider = AuditLoggingProvider.ONLY_ACTOR)
  override fun webBackendUpdateConnection(
    @Body webBackendConnectionUpdate: WebBackendConnectionUpdate,
  ): WebBackendConnectionRead? =
    execute {
      TracingHelper.addConnection(webBackendConnectionUpdate.connectionId)
      webBackendConnectionsHandler.webBackendUpdateConnection(webBackendConnectionUpdate)
    }

  @Post("/connections/mappers/validate")
  @Secured(AuthRoleConstants.WORKSPACE_EDITOR, AuthRoleConstants.ORGANIZATION_EDITOR)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun webBackendValidateMappers(
    @Body webBackendValidateMappersRequestBody: WebBackendValidateMappersRequestBody,
  ): WebBackendValidateMappersResponse? =
    execute {
      webBackendMappersHandler.validateMappers(
        webBackendValidateMappersRequestBody,
      )
    }

  @Post("/describe_cron_expression")
  @Secured(AuthRoleConstants.AUTHENTICATED_USER)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun webBackendDescribeCronExpression(
    @Body body: WebBackendDescribeCronExpressionRequestBody,
  ): WebBackendCronExpressionDescription? =
    execute {
      webBackendCronExpressionHandler.describeCronExpression(
        body,
      )
    }

  @Get("/config")
  @Secured(SecurityRule.IS_ANONYMOUS)
  @ExecuteOn(AirbyteTaskExecutors.IO)
  override fun getWebappConfig(): WebappConfigResponse =
    WebappConfigResponse().apply {
      version = webappConfig.version
      edition = webappConfig.edition.lowercase()
      datadogApplicationId = webappConfig.webApp["datadog-application-id"]
      datadogClientToken = webappConfig.webApp["datadog-client-token"]
      datadogEnv = webappConfig.webApp["datadog-env"]
      datadogService = webappConfig.webApp["datadog-service"]
      datadogSite = webappConfig.webApp["datadog-site"]
      hockeystackApiKey = webappConfig.webApp["hockeystick-api-key"]
      launchdarklyKey = webappConfig.webApp["launchdarkly-key"]
      osanoKey = webappConfig.webApp["osano-key"]
      segmentToken = webappConfig.webApp["segment-token"]
      zendeskKey = webappConfig.webApp["zendesk-key"]
    }
}

/**
 * This class is populated by Micronaut with the values from the `airbyte.web-app` section in the application.yaml.
 *
 * It is only used for by [WebBackendApiController.getWebappConfig].
 *
 * This class should be internal, but due to airbyte-server-wrapped extending the [WebBackendApiController] class, this must be public.
 */
@ConfigurationProperties("airbyte")
data class WebappConfig(
  @Value("\${AIRBYTE_VERSION}") val version: String,
  @Value("\${AIRBYTE_EDITION}") val edition: String,
  val webApp: Map<String, String>,
)
